compiler/pipes/collect-forkable-types.h

```diff
@@ -0,0 +1,41 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2020 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include "common/mixin/not_copyable.h"
+#include "common/smart_ptrs/singleton.h"
+#include "compiler/function-pass.h"
+
+class CollectForkableTypes final : public FunctionPassBase {
+public:
+  string get_description() final {
+    return "Collect forkable types";
+  }
+
+  VertexPtr on_enter_vertex(VertexPtr root) final;
+
+  void on_finish() final;
+
+private:
+  std::vector<const TypeData *> waitable_types_;
+  std::vector<const TypeData *> forkable_types_;
+};
+
+class ForkableTypes : vk::not_copyable {
```

Comment:
Why not to store these fields in CollectForkableTypesPass? 
For example, name waitable_types_ above to waitable_types_current_, and waitable_types_ as waitable_types_total? 
(It seems, that it's quite unintuitive, that a separate class represents the total result of a pass)

Reply:
CollectForkableTypesPass objects have limited lifetime, so if I want to store elements there I have to use static fields, so I've chosen singleton object.

Reply:
Yes, sure, I meant of course to use static fields. Do you just prefer a singleton object over using static fields in a pass, or there is another reason?

Reply:
Yes, I prefer singleton, so I can split these pipe and storage logic, noting more.